### PR TITLE
Multi project extension

### PR DESCRIPTION
**Pull request by MysticJay**
This is the first and IMHO most important plugin ZasoGD has created.
The script itself is unchanged compared to the original file.
Headers and footers are changed (not sure about the `@match` line)

Zaso's startup wrapper has been commented out, although I do not yet understand why.

The script has been tested in IITC-CE - Menu item load.

### CODE REVIEW

File: plugins/multi-projects-extension.user.js [hunk 1]
```diff
@@ -0,0 +1,544 @@
+// ==UserScript==
+// @id             iitc-plugin-multi-projects-extension@Zaso
+// @name           IITC plugin: Multi Projects Extension
+// @category       Control
+// @version        0.0.6.@@DATETIMEVERSION@@
+// @namespace      http://www.giacintogarcea.com/ingress/items/
```
Comment (modos189) [inline plugins/multi-projects-extension.user.js]: Please remove lines @namespace, @updateURL, @downloadURL, @match and @grand

File: plugins/multi-projects-extension.user.js [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+// @updateURL      http://www.giacintogarcea.com/ingress/iitc/multi-projects-extension-by-zaso.meta.js
+// @downloadURL    http://www.giacintogarcea.com/ingress/iitc/multi-projects-extension-by-zaso.user.js
+// @description    [@@BUILDNAME@@-@@BUILDDATE@@] Create separated projects in some plugins.
+// @match          https://intel.ingress.com/*
+// @grant          none
+@@METAINFO@@
+// ==/UserScript==
+@@PLUGINSTART@@
+// function wrapper(){
```
Comment (modos189) [inline plugins/multi-projects-extension.user.js]: The wrapper is added when building the project, please remove these comments

File: plugins/multi-projects-extension.user.js [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+		window.plugin.mpe.setupCSS();
+		window.plugin.mpe.ui.addControl();
+		window.plugin.mpe.ui.appendContainerInSidebar();
+	}
+
+// PLUGIN END //////////////////////////////////////////////////////////
+
+@@PLUGINEND@@
+
+// if(window.iitcLoaded && typeof setup === 'function'){
```
Comment (modos189) [inline plugins/multi-projects-extension.user.js]: and this comments

**Comment by MysticJay** [comment]
IMHO ready to merge

File: plugins/multi-projects-extension.user.js [hunk 4]
```diff
@@ -0,0 +1,544 @@
...
+// ==/UserScript==
+@@PLUGINSTART@@
+// function wrapper(){
+// // ensure plugin framework is there, even if iitc is not yet loaded
+// if(typeof window.plugin !== 'function') window.plugin = function(){};
+
+// PLUGIN START ////////////////////////////////////////////////////////
+
+	// use own namespace for plugin
+	window.plugin.mpe = function() {};
```
Comment (modos189) [inline plugins/multi-projects-extension.user.js]:
```suggestion
window.plugin.mpe = function() {};
```

Comment (modos189) [reply plugins/multi-projects-extension.user.js]: Reduce the indent for all code.

Everything else is fine.

**Comment by modos189** [comment]
Contents of `Multi-Projects-Extension.doc` file please place as comment in multi-projects-extension.user.js file, and delete Multi-Projects-Extension.doc

**Comment by MysticJay** [comment]
Does IITC-CE work with Block-Comments? /* and */?

**Comment by modos189** [comment]
You can use multiline comments. Yeah, like you just did.

We have documentation (in the doc folder) but I don't like it very much, it should be generated based on comments in the code.

**Comment by modos189** [comment]
Pay attention to my two remaining comments https://github.com/IITC-CE/ingress-intel-total-conversion/pull/221/files

**Review by modos189 — DISMISSED** [review]

**Comment by modos189** [comment]
Thanks, this PR will be merged into a new release

**Comment by johnd0e** [comment]
> Thanks, this PR will be merged into a new release

I'd rather think about include support of this in IITC core.
May be in some other form (need to investigate functionality deeper).

**Comment by MysticJay** [comment]
To test the functionality:
1) import Multi-Project-Extension.user.js to externals
2) import DRAW2-beta.user.js to externals
3) disable regular DRAW-Tools
4) enable MPE and Draw2-beta
Create a drawing
from MultiProjects Menu click on the "+" and create a new project. 
Existing drawing should be cleared.
create a new draw.
Use the dialogue to switch between drawings
Use the settings dialogue to add the selection menu to the MainMenu.

**Comment by MysticJay** [comment]
To test:
disable standard Bookmarks-Plugin,
import bkmks2.beta to externals

**Comment by johnd0e** [comment]
1. I do not get why you need separate plugin (draw2). You may just extend single draw-tools (in runtime)

2. But if you still need draw2 then please make your changes reviewable:
   - copy `draw-tools.user.js` to `draw-tools2.user.js`
   - make your changes in separate commit(s)

**Comment by MysticJay** [comment]
Bare with me, I am new to GIT, so it seemed to be safer to create a separate draw than to fiddle around in the original draw. At least this way does not break the existing tools and existing extensions to it. When I started porting Zaso's changes it was not clear to me how deep the changes would be.

The changes can easily be followed by just renaming DRAW2-beta.user.js to draw-tools.user.js
Also the contributions should show the changes step by step, as the first contribution of the file was the original draw-tools (just a copy)

Draw2 and Bkmarks2 is "only" a Draft to show how MPE is working. Due to the same namespace both use with the old Versions, the data is completely compatible. If you find a problem with MPE it is easy to switch back to the old plugins without loosing anything.

Also there are many other proposals that need to be reviewed and I think it is easier to add these to a forked plugin and finally just replace the old plugin than to daily send in new contributions.

As you wrote earlier we should rather think of adding MPE to the code as "Projects" or "Project-Manager".

**Comment by johnd0e** [comment]
> Also the contributions should show the changes step by step, as the first contribution of the file was the original draw-tools (just a copy)

Yeah, but for `draw-tools` I see only single commit https://github.com/IITC-CE/ingress-intel-total-conversion/pull/221/commits/9d89089e8f3b14fada5be8e1e25543940d69297a

Better if there were (1) initial (2) changes for MPE

> Also there are many other proposals that need to be reviewed and I think it is easier to add these to a forked plugin

We can play with forked plugins, but I do not think that we should release them as they are. It'd be confusing.

> As you wrote earlier we should rather think of adding MPE to the code as "Projects" or "Project-Manager".

Yes. It is general principle.
But there may be different implementations of Multi-Projects idea.
I need time to check it.

**Comment by MysticJay** [comment]
Changes applied on original files. removed development (beta) files.

**Comment by johnd0e** [comment]
Ok. I will review proposed changes ASAP (but do not expect it to be fast, definitely not earlier than 0.30 release).

**Comment by johnd0e** [comment]
Some general thoughts:

MPE allows to group Draws/Bookmarks by related project.
What if we need some particular set of Draws or Bookmarks in several projects?
Oops. We can't.

I suppose we could improve project idea.
E.g.:
Bookmarks already has folders. So we might just select existent folders to add them into project.

File: plugins/font-awesome.user.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
+// ==UserScript==
+// @id             iitc-plugin-font-awesome@Zaso
+// @name           IITC plugin: Font Awesome
```
Comment (johnd0e) [inline plugins/font-awesome.user.js:3]: If we include web font - we should include it in core, to make it available out-of-the-box.
(and it is still unclear if should include - #236)

File: plugins/font-awesome.user.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+// PLUGIN START ////////////////////////////////////////////////////////
+
+window.plugin.faIcon = {};
+
+var setup = function(){
+	$('<link>')
+		.prop('type', 'text/css')
+		.prop('rel', 'stylesheet')
+		.prop('href', 'https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css')
```
Comment (johnd0e) [inline plugins/font-awesome.user.js:19]: Why use 4.7 instead of current (5.10.1)?
https://fontawesome.com/

Comment (MysticJay) [reply plugins/font-awesome.user.js:19]: No objection to use 5.10.1.
Just not tested it.

**Comment by MysticJay** [comment]
MPE is about replacing the complete namespace of a plugin. Sharing parts between different namespaces would require them to be in an additional namespace.
Whatever I do not think this is needed Copy paste between different projects is possible and helps to avoid accidental sharing of confidential Information between projects.
I highly recommend to not share anything between projects.

**Comment by johnd0e** [comment]
I still don't get what do you mean. Explain please.

**Comment by MysticJay** [comment]
You asked about sharing parts of draw or bookmarks between projects.

**Comment by MysticJay** [comment]
What about moving this to the core?

**Comment by johnd0e** [comment]
It is planned

**Pull request closed without merging**